refactor: Make CAddrMan::cs non-recursive

hebasto opened the pull request:
This PR replaces `RecursiveMutex CAddrMan::cs` with `Mutex CAddrMan::cs`.

All of the related code branches are covered by appropriate lock assertions to insure that the mutex locking policy has not been changed by accident.

Related to #19303.

Based on #22025, and first three commits belong to it.

vasild commented on src/addrman.h:
```diff
@@ -534,137 +553,133 @@ friend class CAddrManTest;
     }
 
     //! Return the number of (unique) addresses in all tables.
-    size_t size() const
+    size_t size() const EXCLUSIVE_LOCKS_REQUIRED(!m_addrman_mutex)
     {
-        LOCK(cs); // TODO: Cache this in an atomic to avoid this overhead
-        return vRandom.size();
+        LOCK(m_addrman_mutex); // TODO: Cache this in an atomic to avoid this overhead
+        return sizeNonLockerHelper();
```
It would be nice to implement that TODO comment because it would make it possible to avoid any mutex requirements from `size()`.

I did so in https://github.com/vasild/bitcoin/commit/f25d8a7c6de0e45b037f91138e1b4a72e7a1653f. If you like it you can replace the commit `refactor: Prevent double lock in CAddrMan::size()` with that one. In this branch https://github.com/vasild/bitcoin/tree/200610-addrman-mx that is done and also the subsequent commits are adjusted.

Or, if you don't like the caching, then maybe just delete the TODO comment.

hebasto replied:
The commit you [suggested](https://github.com/vasild/bitcoin/commit/f25d8a7c6de0e45b037f91138e1b4a72e7a1653f) does change `CAddrMan::size()` behavior. So I'd prefer to keep this PR as clean refactoring with no behavior changes.

vasild replied:
ok

vasild commented on src/sync.h:
```diff
@@ -103,6 +103,9 @@ class LOCKABLE AnnotatedMixin : public PARENT
     }
 
     using UniqueLock = std::unique_lock<PARENT>;
+#ifdef __clang__
+    const AnnotatedMixin& operator!() const { return *this; }
+#endif // __clang__
```
Here and in `StdMutex` - I find this very confusing because it would end up with `m` being the same as `!m`. It is a huge hack imposed by clang. I would suggest to add some comment to explain what is that. Something like `Required by the clang thread safety instrumentation so that we can declare that a mutex should not be held when a function is called - EXCLUSIVE_LOCKS_REQUIRED(!m) (aka "negative capabilities")`.

The example in https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#mutex-h also has some comment.

PS I wonder what would `EXCLUSIVE_LOCKS_REQUIRED(!!m)` do. :eyes:

hebasto replied:
Done in #19249.

vasild reviewed (commented):
Looks good, some suggestions below.

The last two commits:
```
Add means to handle negative capabilities in thread safety annotations
refactor: Add negative thread safety annotations to CAddrMan
```
are somewhat unrelated and maybe deserve a separate PR if they cause some controversy or if another reviewer asks to move them away to ease review. Since I already reviewed everything I am fine with them as they are.

hebasto commented:
Updated d9572933f1fb0201b259d9cb7b40873bdbad7b6f -> b6712ece8ada3b1c9230f1264da81cbf11fe4595 ([pr19238.02](https://github.com/hebasto/bitcoin/commits/pr19238.02) -> [pr19238.04](https://github.com/hebasto/bitcoin/commits/pr19238.04)):

- the commit "Add means to handle negative capabilities in thread safety annotations" has been separated into #19249, and this PR has been rebased on the latter one
- negative mutex requirements added to each commit opportunistically

vasild reviewed (approved):
ACK b6712ece8

hebasto commented:
Updated b6712ece8ada3b1c9230f1264da81cbf11fe4595 -> 9cd0ed858c96b1052580e2044069c5cb9434628e ([pr19238.04](https://github.com/hebasto/bitcoin/commits/pr19238.04) -> [pr19238.05](https://github.com/hebasto/bitcoin/commits/pr19238.05)):

- rebased due to the merging of #19249
- all of the related code branches are covered by appropriate lock assertions to insure that the locking policy has not been changed by accident

hebasto commented:
The OP has been updated.

hebasto commented on src/addrman.h:
```diff
@@ -534,137 +559,145 @@ friend class CAddrManTest;
     {
-        LOCK(cs);
-        Check();
+        AssertLockNotHeld(m_addrman_mutex);
+        LOCK(m_addrman_mutex);
```
> Following that, should we add `AssertLockNotHeld()` before every `LOCK()` in the source code?

Only for `Mutex` instances, not for `RecursiveMutex` ones.

> What about adding `AssertLockNotHeld()` inside `LOCK()`?

Not now because the `RecursiveMutex` instance could be an argument of `LOCK()`.

hebasto replied:
> What about checking, inside `LOCK()`, whether the mutex being locked is instance of `Mutex` (and not `RecursiveMutex`) and only asserting that we don't own it in that case

It is not thread safe, for sure.

hebasto replied:
https://softwareengineering.stackexchange.com/questions/330769/why-cant-i-check-whether-a-mutex-is-locked

hebasto replied:
@vasild Mind submitting a PR that it could be reviewed before this one?

hebasto replied:
Currently, it is assumed the following patterns:
- for public interface methods:
```
# foo.h
class Foo
{
public: void foo() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex);
}
# foo.cpp
void Foo::foo()
{
    AssertLockNotHeld(m_mutex);
    LOCK(m_mutex);
    // do magic
}
```

- for internal methods that are called from critical sections:
```
# foo.h
class Foo
{
pivate: void foo_cs() EXCLUSIVE_LOCKS_REQUIRED(m_mutex);
}
# foo.cpp
void Foo::foo_cs()
{
    AssertLockHeld(m_mutex);
    // do magic
}
```

The common places are thread safety annotations and lock assertions. It seems placing `AssertLockNotHeld()` into internals of the `LOCK()` breaks uniformity, no?

vasild replied:
> Mind submitting a PR that it could be reviewed before this one?

Yes, I think that deserves a separate PR. I have to consider this a little bit before opening a PR. Maybe it would be possible to optimize it - the patch above would lock `lockdata.dd_mutex` and dig into the stack of locks for the current thread 2 times - once from the newly added `AssertLockNotHeld()` and a second time from `push_lock()`. It should be possible to do everything in `push_lock()` and avoid adding the call to `AssertLockNotHeld()`.

> ...breaks uniformity, no?

I don't see it that way - `LOCK()` already implies that we don't own the lock.

vasild replied:
Here is a PR that would detect a double lock. With the check from inside `LOCK()` it makes unnecessary to add `AssertLockNotHeld()` calls before each `LOCK()`: https://github.com/bitcoin/bitcoin/pull/19337

hebasto commented:
@sipa Mind reviewing this PR?

hebasto commented:
btw, how about convention to add the `_cs` suffix to the name of the function that is called from within the critical section already guarded by the locked instance of the `Mutex`?

vasild commented:
> btw, how about convention to add the `_cs` suffix to the name of the function that is called from within the critical section already guarded by the locked instance of the `Mutex`?

I like it, and it is shorter than `...NonLockHelper()`, but what about the [CamelCaseConvention](
https://github.com/bitcoin/bitcoin/blame/808ef36b89ea9ce72116bbd7ee479b984367dc60/doc/developer-notes.md#L87)? Also, it would get fuzzy if more than one mutex is involved and a method requires one to be held and the other not...

hebasto commented:
Note for reviewers: please review #19337 at fist.

hebasto commented:
Updated 9cd0ed858c96b1052580e2044069c5cb9434628e -> d985aeeac48114f1078980888303f34ffdb78e74 ([pr19238.05](https://github.com/hebasto/bitcoin/commits/pr19238.05) -> [pr19238.06](https://github.com/hebasto/bitcoin/commits/pr19238.06)):

- rebased due to the conflict with #19658
- implemented naming [convention](https://github.com/bitcoin/bitcoin/pull/19238#issuecomment-646226615) with :+1:  from @ajtowns 
- separated commit with potentially controversial mutex renaming

jnewbery commented:
Concept ACK. Needs rebase.

hebasto commented:
Updated d985aeeac48114f1078980888303f34ffdb78e74 -> f1246cb12f573c4a8b32d75d6ea377eeec3b43e7 ([pr19238.06](https://github.com/hebasto/bitcoin/commits/pr19238.06) -> [pr19238.07](https://github.com/hebasto/bitcoin/commits/pr19238.07)):

- rebased
- dropped the mutex renaming commit

jnewbery commented on src/addrman.h:
```diff
@@ -262,7 +297,7 @@ friend class CAddrManTest;
 
     //! Mark an entry as attempted to connect.
-    void Attempt_(const CService &addr, bool fCountFailure, int64_t nTime) EXCLUSIVE_LOCKS_REQUIRED(cs);
+    void AttemptWithLockHeld(const CService &addr, bool fCountFailure, int64_t nTime) EXCLUSIVE_LOCKS_REQUIRED(cs);
```
I don't think we need to expose this as a public method just for the addrman tests. Just release the lock in that test before calling `Attempt()`:

```diff
--- a/src/test/addrman_tests.cpp
+++ b/src/test/addrman_tests.cpp
@@ -73,13 +73,12 @@ public:
     // Simulates connection failure so that we can test eviction of offline nodes
     void SimConnFail(const CService& addr)
     {
-         LOCK(cs);
-         int64_t nLastSuccess = 1;
-         Good_(addr, true, nLastSuccess); // Set last good connection in the deep past.
+         // Set last good connection in the deep past.
+         WITH_LOCK(cs, {Good_(addr, true, /* nLastSuccess= */ 1);});
 
          bool count_failure = false;
          int64_t nLastTry = GetAdjustedTime()-61;
-         AttemptWithLockHeld(addr, count_failure, nLastTry);
+         Attempt(addr, count_failure, nLastTry);
```

The unit test is single threaded, so there's no need to hold the mutex between `Good_()` and `Attempt()`.

hebasto replied:
Thanks! Done in bc3694a49a93b9ae6f8cacfd0d5b1def5f09b004.

jnewbery commented on src/addrman.h:
```diff
@@ -229,13 +229,48 @@ friend class CAddrManTest;
     //! Holds addrs inserted into tried table that collide with existing entries. Test-before-evict discipline used to resolve these collisions.
     std::set<int> m_tried_collisions;
 
+    void CheckWithLockHeld() EXCLUSIVE_LOCKS_REQUIRED(cs)
```
Make this (and other `Check()` functions) `const`?

hebasto replied:
This change is not trivial due to the non-const [`std::map::operator[]`](https://en.cppreference.com/w/cpp/container/map/operator_at):
```
addrman.cpp:433:20: error: no viable overloaded operator[] for type 'const std::map<CNetAddr, int>'
        if (mapAddr[info] != n)
            ~~~~~~~^~~~~
```

jnewbery commented on src/addrman.h:
```diff
@@ -229,13 +229,48 @@ friend class CAddrManTest;
+#endif // DEBUG_ADDRMAN
+    }
+
+    void ClearWithLockHeld() EXCLUSIVE_LOCKS_REQUIRED(cs)
```
It'd be good to get rid of the `Clear()` functions entirely:

- the call to `Clear()` in deserialize could be replaced with an assert that size() is 0 (we should never deserialize into a populated `CAddrman`
- the calls in the tests can be replaced by instantiating new `CAddrman`s. Clearing then reusing an existing `CAddrman` is not a realistic usage pattern.

If you don't want to do that in this PR, it feels like this function should be called `Clear_()` and be defined in the cpp file, for consistency with the other functions.

hebasto replied:
@jnewbery 
> It'd be good to get rid of the `Clear()` functions entirely:

I totally agree with you.

>  - the call to `Clear()` in deserialize could be replaced with an assert that size() is 0 (we should never deserialize into a populated `CAddrman`)

Going to implement this suggestion.

>  - the calls in the tests can be replaced by instantiating new `CAddrman`s. Clearing then reusing an existing `CAddrman` is not a realistic usage pattern.

This part is not trivial, and probably deserves its own pr.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/19238#issuecomment-846605462).

jnewbery commented on src/addrman.h:
```diff
@@ -679,9 +692,9 @@ friend class CAddrManTest;
         CAddrInfo ret;
```
It feels like this is a good opportunity to make the access patterns consistent here:

```diff
@@ -689,13 +689,10 @@ public:
     //! Randomly select an address in tried that another address is attempting to evict.
     CAddrInfo SelectTriedCollision()
     {
-        CAddrInfo ret;
-        {
-            LOCK(cs);
-            CheckWithLockHeld();
-            ret = SelectTriedCollision_();
-            CheckWithLockHeld();
-        }
+        LOCK(cs);
+        CheckWithLockHeld();
+        CAddrInfo ret = SelectTriedCollision_();
+        CheckWithLockHeld();
         return ret;
     }
 
@@ -704,26 +701,21 @@ public:
      */
     CAddrInfo Select(bool newOnly = false)
     {
-        CAddrInfo addrRet;
-        {
-            LOCK(cs);
-            CheckWithLockHeld();
-            addrRet = Select_(newOnly);
-            CheckWithLockHeld();
-        }
+        LOCK(cs);
+        CheckWithLockHeld();
+        CAddrInfo addrRet = Select_(newOnly);
+        CheckWithLockHeld();
         return addrRet;
     }
 
     //! Return a bunch of addresses, selected at random.
     std::vector<CAddress> GetAddr(size_t max_addresses, size_t max_pct)
     {
-        Check();
+        LOCK(cs);
+        CheckWithLockHeld();
         std::vector<CAddress> vAddr;
-        {
-            LOCK(cs);
-            GetAddr_(vAddr, max_addresses, max_pct);
-        }
-        Check();
+        GetAddr_(vAddr, max_addresses, max_pct);
+        CheckWithLockHeld();
         return vAddr;
     }
```

(i.e. for all of these functions take the lock, run Check(), run the inner function, run Check() again, release lock/return)

vasild replied:
:+1: This would make the `Check()` function unused, so it can be removed.

hebasto replied:
Thanks! Done in e21c0dd57009197dfeedbc4ed4d30454b7a05cf6.

hebasto commented:
@jnewbery Thank you for your review. While working on this pull I had similar thoughts about points you mentioned. So I'm happy to make some improvements.

jnewbery commented:
@hebasto Whilst you're looking at this code, do you mind taking a look at #20233. Without that PR, `CAddrMan::Check()` is basically useless, since it's enabled by a compile-time option which I expect nobody uses. That PR would make it a runtime check, so that we could enable it for the unit, functional and fuzz tests. It's obviously much more likely to catch bugs if it's being run by those tests!

hebasto commented:
> @hebasto Whilst you're looking at this code, do you mind taking a look at #20233. Without that PR, `CAddrMan::Check()` is basically useless, since it's enabled by a compile-time option which I expect nobody uses. That PR would make it a runtime check, so that we could enable it for the unit, functional and fuzz tests. It's obviously much more likely to catch bugs if it's being run by those tests!

Sure!

ajtowns commented on src/addrman.h:
```diff
@@ -608,23 +625,19 @@ friend class CAddrManTest;
     void Check()
```
I was expecting to see `EXCLUSIVE_LOCKS_REQUIRED(!cs)` added to these functions that lock `cs` to get compile time detection of double-locking. Any reason not to have them?

hebasto replied:
Of course! Done in 7f01942d2d3d9d66a8932d6a1fe08281d9b52273.

vasild commented:
0c0fb742ad7b686209efb627a8ec43d7f96c3754 looks good, I agree with the above suggestions from @jnewbery and @ajtowns.

vasild commented on src/addrman.h:
```diff
@@ -229,13 +229,48 @@ friend class CAddrManTest;
     //! Source of random numbers for randomization in inner loops
     FastRandomContext insecure_rand;
 
+private:
```
Yeah, class declaration is easier to read if there is just one instance of `public:`, `protected:` and `private:` (in that order).

hebasto replied:
@jnewbery @vasild 

Done in #22025.

hebasto committed:
test: Drop excessive locking in CAddrManTest::SimConnFail

The unit test is single threaded, so there's no need to hold the mutex
between Good() and Attempt().

This change avoids recursive locking in the CAddrMan::Attempt function.

Co-authored-by: John Newbery <john@johnnewbery.com>

hebasto committed:
refactor: Avoid recursive locking in CAddrMan::size

hebasto committed:
Make CAddrMan::Check private

Change in the addrman.h header is move-only.

hebasto committed:
refactor: Fix CAddrMan::Check style

This change improves readability, and follows Developer Notes.

hebasto committed:
refactor: Avoid recursive locking in CAddrMan::Check

hebasto committed:
refactor: Apply consistent pattern for CAddrMan::Check usage

Co-authored-by: John Newbery <john@johnnewbery.com>

hebasto committed:
refactor: Avoid recursive locking in CAddrMan::Clear

Co-authored-by: John Newbery <john@johnnewbery.com>

hebasto committed:
Add thread safety annotations to CAddrMan public functions

hebasto committed:
Add AssertLockHeld to CAddrMan private functions

hebasto committed:
refactor: Make CAddrMan::cs non-recursive

hebasto commented:
@jnewbery @ajtowns @vasild 

Thank you for your reviews!

Updated 0c0fb742ad7b686209efb627a8ec43d7f96c3754 -> 148a7c9454adc3e311867f3566e01d261b362f71 ([pr19238.08](https://github.com/hebasto/bitcoin/commits/pr19238.08) -> [pr19238.09](https://github.com/hebasto/bitcoin/commits/pr19238.09)):

- addressed comments
- rebased on top of the recent CI changes

vasild reviewed (approved):
ACK 148a7c9454adc3e311867f3566e01d261b362f71

This PR includes https://github.com/bitcoin/bitcoin/pull/22025.

Compiles without warnings with Clang 13.0.0.

hebasto commented on src/addrman.h:
```diff
@@ -752,13 +611,161 @@ friend class CAddrManTest;
+    {
+#ifdef DEBUG_ADDRMAN
+        AssertLockHeld(cs);
+        if (const int err = Check_(); err != 0) {
```
Thanks! Updated.

vasild reviewed (approved):
ACK ee79105070778bd1f3f9693536e3c92801ae63a8

hebasto commented:
Rebased ee79105070778bd1f3f9693536e3c92801ae63a8 -> e2787fb24a675e62a82bcb43400ad8e1a68761d2 ([pr19238.10](https://github.com/hebasto/bitcoin/commits/pr19238.10) -> [pr19238.11](https://github.com/hebasto/bitcoin/commits/pr19238.11)) due to the merging of #22025.

vasild reviewed (approved):
ACK e2787fb24a675e62a82bcb43400ad8e1a68761d2

hebasto commented:
Rebased e2787fb24a675e62a82bcb43400ad8e1a68761d2 -> af9550561d7fa8cfa8da910db3e46993f848a0f5 ([pr19238.11](https://github.com/hebasto/bitcoin/commits/pr19238.11) -> [pr19238.12](https://github.com/hebasto/bitcoin/commits/pr19238.12)) due to the conflict with #18722.

vasild reviewed (approved):
ACK af9550561d7fa8cfa8da910db3e46993f848a0f5

jnewbery commented:
Code review ACK af9550561d7fa8cfa8da910db3e46993f848a0f5

MarcoFalke commented:
There might be a silent merge conflict

MarcoFalke commented:
```
test/fuzz/addrman.cpp:110:30: error: 'Check' is a private member of 'CAddrMan'
    (void)/*const_*/addr_man.Check();
                             ^
./addrman.h:722:10: note: declared private here
    void Check()
         ^
1 error generated.

hebasto commented:
> There might be a silent merge conflict

With #21941. Rebasing.

hebasto commented:
@MarcoFalke @jnewbery @vasild 

Rebased af9550561d7fa8cfa8da910db3e46993f848a0f5 -> ae98aec9c0521cdcec76459c8200bd45ff6a1485 ([pr19238.12](https://github.com/hebasto/bitcoin/commits/pr19238.12) -> [pr19238.13](https://github.com/hebasto/bitcoin/commits/pr19238.13)) due to the silent merge conflict with #21941.

vasild reviewed (approved):
ACK ae98aec9c0521cdcec76459c8200bd45ff6a1485
